sdks/python/apache_beam/tools/utils.py

```diff
@@ -32,3 +40,67 @@ def check_compiled(module):
         "Profiling uncompiled code.\n"
         "To compile beam, run "
         "'pip install Cython; python setup.py build_ext --inplace'")
+
+
+def run_benchmarks(benchmarks, num_runs, print_output):
+  """Runs benchmarks, and collects execution times.
+
+  A simple instrumentation to run any python function several times
+  on a generated input, collect and print its execution times.
+
+  Args:
+    benchmarks: A list of dictionaries that describe benchmarks.
```

Comment:
Perhaps it's being too functional, but you could also let a benchmark be any callable that takes as input an input size, and returns a callable. It's name could be getattr(benchmark, '__name__', str(benchmark)) and the input size would be provided externally. 

One could implement benchmarks, of desired, in OO style as

```
class MyBenchmark(object):
  def __init__(self, size):
    [whatever initialization]
  def __call__(self):
    [run the code in question]
```

This would make it easy to write

def run_benchmarks([list of benchmark classes/callables], [list of sizes]):
  ...

Reply:
Thanks for the feedback, @robertwb !
In this case it sounds like we'd have to create a new instance of MyBenchmark for every iteration. Perhaps it would be better to just go OO style as in

```
class Benchmark(object):
  def __init__(self):
     [whatever one-time initialization, for example instantiate coders and create an impl]

  def run(self):
    raise NotImplementedError()
    [run the code in question]
 
 def generate_next_input(size):
    raise NotImplementedError   

 def __str__():
    raise NotImplementedError
    # returns name of this Benchmark. Will be used as a key.

def run_benchmarks([list of (benchmark object, size, number_of_iterations)] ):
   # Returns dict [name->(list of execution times)] 
...
```
This way it seems that it is easier to separate one time initialization from input generation. Reusing object state may help to reduce the overheads related to benchmark execution. 
What do you think?